### ISSUE TITLE
Retrieve the source of a mouse event

### ISSUE THREAD

**Issue by msrd0** [linebender/druid#1347]
I would like to distinguish between events triggered by a mouse that is intended as a pointer device (like a real mouse, trackpad etc) and events that happen on a touchscreen, tablet etc. However, I could not find anything on the `MouseEvent` that would tell me if it was triggered by a mouse or by a touchscreen. Is it possible to know the device that triggered the event, or to include some more meta information with the event, like its source device type?

**Comment by Finnerale** [comment]
As far as I know, Druid does not support touch input on any Platform. If it happens to work somewhere, then probably only because touch input is simulating a mouse.

**Comment by msrd0** [comment]
Well, it happens to work on wayland, so maybe touch is simulating mouse events. Anyways, is there any possibility to know where the (simulated) event came from? Or maybe it's possible to add one?

**Comment by Finnerale** [comment]
I think that would require actually implementing real touch support in Druid.

**Comment by cmyr** [comment]
I think this will have to wait for actual multi-touch support to be implemented, which isn't current planned. It would definitely be *possible* to add a `DeviceType` enum to druid-shell and include that in `MouseEvent`, but it isn't a priority; if you wanted to write that PR, though, assuming it is relatively non-invasive (which I expect) I'd be happy to merge it on the understanding that it will break at some point (although hopefully only once we've implemented a suitable replacement).

**Comment by msrd0** [comment]
I played with what gtk gives me and it seems that gtk, when emulating touch events as mouse events, also classifies them as comming from a mouse. So we'd probably need to somehow tell gtk that we do want touch events rather than emulated mouse events. If you're interested in merging a linux-only solution I can try to work on a PR.

**Comment by cmyr** [comment]
I don't think we want gtk to give us touch events, because then we'll actually need to handle them and do that conversion ourselves, which will be fiddly; all we would want is some way to query the device that generated the event in question. If we can't get that easily I don't think there's a good answer until we implement real multi-touch. :(

also: it looks like on macOS I can't easily distinguish between, say, a mouse and a touchpad. I assume there's a lower-level API, but there's not much in AppKit; I may be able to determine if a touch comes from a tablet or a 'mouse' (which includes the trackpad) but not much more than that... we'll see if this changes if we get touch-screen macs in the nearish future?

**Comment by msrd0** [comment]
Yeah I agree, handling touch events is probably not desirable. The best thing we could do without is look at the device names, my mouse is called `Core Pointer` whereas my touchscreen is called `Wayland Touch Master Pointer`. But that doesn't sound like a reliable solution ...

**Comment by doppioandante** [comment]
Hello, I'm writing a note-taking application with druid. I have a very small patch to druid in my own fork to handle stylus events.
Now I'm at the point where I would like to explore handling touch events.
My primary interest is linux, so gtk, but I would be glad to start discussing a general druid implementation.
So my idea was to enable TOUCH_MASK in the gdk events and sort of map gtk touch events into druid
> I don't think we want gtk to give us touch events, because then we'll actually need to handle them and do that conversion ourselves, which will be fiddly; 
I fear I don't know enough to understand this, but why would it be undesirable?

**Comment by cmyr** [comment]
@doppioandante My thinking was just that having GTK do conversion is going to just be an extra layer of logic we potentially have to work around in order to get the behaviour we want; if we can deal with events at a lower level we  would have more control over how we surface them.

**Comment by PoignardAzur** [comment]
I'm assuming this is WONTFIX for now, until we implement proper touch events. In the meantime, I'm closing this issue.
